src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc

```diff
@@ -0,0 +1,927 @@
+
+    if (!this->node.Subscribe(topic,
+        &OpticalTactilePluginPrivate::DepthCameraCallback,
+        this))
+    {
+      ignerr << "Error subscribing to topic " << "[" << topic << "]. "
+        << "<topic> must not contain '/'" << std::endl;
+      return;
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::Enable(const bool _value)
+{
+    // Just a placeholder method for the moment
+    _value;
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::FilterOutCollisions(
+                            const EntityComponentManager &_ecm,
+                            const std::vector<Entity> &_entities)
+{
+    if (_entities.empty())
+        return;
+
+    // Clear vector that contains old data
+    this->filteredCollisionEntities.clear();
+
+    // Get collisions from the sensor
+    for (const Entity &entity : _entities)
+    {
+        std::string name = scopedName(entity, _ecm);
+
+        if (name.find(this->modelName) != std::string::npos)
+        {
+            this->filteredCollisionEntities.push_back(entity);
+        }
+    }
+}
+
+//////////////////////////////////////////////////
+void OpticalTactilePluginPrivate::InterpolateData(
+                    const ignition::msgs::Contact &contact)
+{
+    // The Optical Tactile Sensor is supposed to be box-shaped.
+    // Contacts returned by the Contact Sensor in this type of shapes should be
+    // 4 points in the corners of some of its faces, in this case the one
+    // corresponding to the 'sensing' face.
+    // Given that the number of contacts may change even when the objects are
+    // at rest, data is only interpolated when the number of contacts
+    // returned is 4.
+
+    if (contact.position_size() == 4)
+    {
+        ignition::math::Vector3d contact1 =
+            ignition::msgs::Convert(contact.position(0));
+        ignition::math::Vector3d contact2 =
+            ignition::msgs::Convert(contact.position(1));
+        ignition::math::Vector3d contact3 =
+            ignition::msgs::Convert(contact.position(2));
+        ignition::math::Vector3d contact4 =
+            ignition::msgs::Convert(contact.position(3));
+
+        ignition::math::Vector3d direction1 =
+            (contact2 - contact1).Normalized() * this->resolution;
+        ignition::math::Vector3d direction2 =
+            (contact4 - contact1).Normalized() * this->resolution;
+
+        ignition::math::Vector3d interpolatedVector = contact1;
+        this->interpolatedPosition.push_back(contact1);
+
+        // auxiliary Vector3d to iterate through contacts
+        ignition::math::Vector3d tempVector(0.0, 0.0, 0.0);
+
+        uint64_t steps1 =
+            contact1.Distance(contact2) / this->resolution;
+        uint64_t steps2 =
+            contact1.Distance(contact4) / this->resolution;
+
+        for (uint64_t index1 = 0; index1 <= steps1; ++index1)
```

Comment:
> Unconventional practice to use 0 with `<=`

Just for curiosity, should it be `0` with `<` and `1` with `<=`?

> Right now, the loop produces (steps1+1) * steps2 iterations.

Error there

Reply:
Yes usually `0` with `<` is the preference (unless you're in MATLAB, which is......... another story).

Reply:
Ok, thanks!

> Error there

I think I was too quick saying that. The for loop actually computes `(steps1+1) * (steps2+1)` points that way. If you take for instance a 3x3 cm with a 1 cm resolution you get 4x4 points that way. I do `index1 <= steps1` because otherwise I don't get the last column. Does that makes sense?

Reply:
Does it compute steps2+1 in the inner loop? That was my original question. Currently, it looks like it's only computing steps2 iterations, in `[0, steps2)`. The outer loop is computing `[0, steps1]`. So the total right now is (steps1+1) * steps2, I think.

Reply:
It computes `steps2` in the inner loop and `steps1 + 1` in the outer loop, but the thing is that the total number of steps is not the same that the total number of points `interpolatedVector` passes through

Reply:
So to clarify, why does the inner loop only needs `steps2` iterations? Does it not also need `steps2 + 1`, to get the last row?

Reply:
No, it doesn't.

Reply:
Because one of the rows is already being got [here](https://github.com/ignitionrobotics/ign-gazebo/blob/16732f6f7b7a1e36f743fed07458296bafa177c1/src/systems/optical_tactile_plugin/OpticalTactilePlugin.cc#L409). The reason of that if-block statement is to not make `interpolatedVector` go outside of the rectangle drawn by the contacts' data.

Reply:
Hmm I think a few comments in this loop, at least a comment at the top of this loop noting what the indexing is doing, and changing the variable names `1` and `2` to something referring to row and column like `r` and `c`, would make this method more readable. Or is it not known whether it's row or column because the directions might change?
Are the 4 points returned in clockwise / counterclockwise order? A comment on that where contacts 1-4 are declared also helps the reader.

I just realized that this is a method that I said elsewhere that we should remove / comment out? I can't find that spot anymore. Is this that method? If we're going to remove it anyway, then we don't need to spend time on it.

Reply:
> I just realized that this is a method that I said elsewhere that we should remove / comment out? I can't find that spot anymore. Is this that method? If we're going to remove it anyway, then we don't need to spend time on it.

Yeah, here's the comment https://github.com/ignitionrobotics/ign-gazebo/pull/229#discussion_r466915015

Reply:
Yeah I saw that, but the line number and method name there didn't match with this version, so I was confused. Now I see that that comment is from the 20th commit, and this set is only the first 19, so that comment wasn't showing here.

So here are the two options. I would say just remove this method since we're not going to use it anymore. You can always retrieve it from Git history if you need it later. If you really want to keep it now, then it should be tidied up with additional comments as mentioned above, and then the method call commented out.

Reply:
Removed the method in e216bb2